### PR TITLE
create backend

### PR DESCRIPTION
**Pull request by gevondyanvaagn**

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
+const express = require('express');
+const bodyParser = require('body-parser');
+const cors = require('cors');
+const { default: mongoose } = require('mongoose');
```
Comment (baklanovairina) [inline app.js:4]: Where is this entry from?

File: app.js [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+const app = express();
+
+const uri = "mongodb+srv://gev:Gevondyan123@cluster0.afwpn.mongodb.net/user?retryWrites=true&w=majority"
```
Comment (baklanovairina) [inline app.js:7]: Install "dotenv", create a ".env" file and put your database address there. And then your data will be in the public domain.

File: app.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
+const express = require('express');
+const bodyParser = require('body-parser');
```
Comment (baklanovairina) [inline app.js:2]: It's outdated

File: app.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+const app = express();
+
+const uri = "mongodb+srv://gev:Gevondyan123@cluster0.afwpn.mongodb.net/user?retryWrites=true&w=majority"
+mongoose.connect(uri, { useNewUrlParser: true, useUnifiedTopology: true });
+
+const apiRoutes = require("./src/modules/routes/routes");
+
+app.use(cors());
+
+app.use(bodyParser.json());
```
Comment (baklanovairina) [inline app.js:14]: use express instead of bodyParser

File: app.js [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+app.use("/", apiRoutes);
+
+app.listen(8000, () => {
+    console.log('Examp app listening on port 8000')
```
Comment (baklanovairina) [inline app.js:18]: check tab

File: app.js [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+});
```
Comment (baklanovairina) [inline app.js:19]: create function and take out there connection to a database and connection to port. And then at the end of this file call this function

File: src/db/models/task/index.js [hunk 7]
```diff
@@ -0,0 +1,10 @@
+const mongoose = require('mongoose');
+const {Schema} = mongoose;
```
Comment (baklanovairina) [inline src/db/models/task/index.js:2]: add spaces

File: src/db/models/task/index.js [hunk 8]
```diff
@@ -0,0 +1,10 @@
...
+//const Task = mongoose.model("tasks", taskScheme);
```
Comment (baklanovairina) [inline src/db/models/task/index.js:3]: delete comment

File: src/db/models/task/index.js [hunk 9]
```diff
@@ -0,0 +1,10 @@
...
+
+const taskScheme = new Schema ({
+    text:String,
+    isCheck: Boolean,
+});
+
+module.exports = Task = mongoose.model("tasks",taskScheme);
```
Comment (baklanovairina) [inline src/db/models/task/index.js:10]: "Tasks"

File: src/modules/routes/routes.js [hunk 10]
```diff
@@ -0,0 +1,23 @@
+const express = require('express');
```
Comment (baklanovairina) [inline src/modules/routes/routes.js:1]: The file with routes should contain only routes. And the controllers should be in a separate folder and file

File: src/modules/routes/routes.js [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+const Task = require("../../db/models/task/index");
+
+const router = express.Router();
+
+
+router.post('/creatTask', (req, res) => {
+  const task = new Task({
+    text: req.query.name,
```
Comment (baklanovairina) [inline src/modules/routes/routes.js:9]: Why are you passing text through query and not through body ?

File: src/modules/routes/routes.js [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+    res.send(result);
+  })
+});
+
+router.get('/allTasks', (req, res) => {
+  Task.find().then(result => {
+    res.send({ data: result });
+  })
+
+});
```
Comment (baklanovairina) [inline src/modules/routes/routes.js:22]: where are the routes??

File: src/modules/controllers/task.controller.js [hunk 13]
```diff
@@ -0,0 +1,21 @@
+const Tasks = require("../../db/models/task/index");
```
Comment (baklanovairina) [inline src/modules/controllers/task.controller.js:1]: Delete the db and task folders, and change the name of the controller file. To make it clear what these controllers are for

File: src/modules/controllers/task.controller.js [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+
+module.exports.getAllTasks = async (req, res) => {
```
Comment (baklanovairina) [inline src/modules/controllers/task.controller.js:3]: We do not export functions separately, we do it at the end of the file and export all functions at once

File: src/modules/controllers/task.controller.js [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+  Tasks.find().then(result => {
```
Comment (baklanovairina) [inline src/modules/controllers/task.controller.js:4]: use async/await and add try/catch

File: src/modules/controllers/task.controller.js [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+  })
+};
+
+module.exports.createNewTask = async (req, res) => {
+  const body = req.body
+  const task = new Tasks({
+    text: body.text,
+    isCheck: body.isCheck,
+  });
+  task.save();
```
Comment (baklanovairina) [inline src/modules/controllers/task.controller.js:15]: What are you sending back to the front?

File: src/modules/controllers/task.controller.js [hunk 17]
```diff
@@ -0,0 +1,21 @@
...
+};
+module.exports.deleteTask = (req, res, next) => {
+  const tasksId = req.query.id;
+  Tasks.deleteMany({ _id: tasksId }, function (err, result) {
```
Comment (baklanovairina) [inline src/modules/controllers/task.controller.js:19]: also something unclear.
What are you sending back to the front?

File: src/modules/routes/routes.js [hunk 18]
```diff
@@ -0,0 +1,16 @@
...
+const {
+  getAllTasks,
+  createNewTask,
+  changeTaskInfo,
+  deleteTask
+} = require('../controllers/task.controller');
+
+router.get('/allTasks', getAllTasks);
+router.post('/createTask', createNewTask);
+router.delete('/deleteTask', deleteTask);
```
Comment (baklanovairina) [inline src/modules/routes/routes.js:14]: where is the patch?

File: src/modules/controllers/task.controller.js [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+    res.send({ data: result });
+  } catch (err) {
+    res.status(400).send("error");
+  }
+};
+
+const createNewTask = async (req, res) => {
+  try {
+    const text = req.body.text;
+    if (req.body.hasOwnProperty('text') && text !== '') {
```
Comment (spyatnicyn) [inline src/modules/controllers/task.controller.js:14]: Lets use middlewares to validate request body/query/params. You can use for it libraries like `Joi` or `express-validator`

File: .gitignore [hunk 20]
```diff
@@ -0,0 +1,2 @@
+/node_modules
+/.env
```
[the remark this answers is not in the document]
Comment (spyatnicyn) [reply .gitignore:2]: maybe my previous comment was wrong

File: src/modules/routes/task.js [hunk 21]
```diff
@@ -0,0 +1,21 @@
+const express = require('express');
+const router = express.Router();
+const { body, validationResult } = require('express-validator');
```
Comment (spyatnicyn) [inline src/modules/routes/task.js:3]: validationResult has been imported but not used

File: src/modules/controllers/task.controller.js [hunk 22]
```diff
@@ -0,0 +1,97 @@
...
+  } catch (err) {
+    res.status(400).send("error");
+  }
+};
+
+const changeTaskInfo = async (req, res, next) => {
+  try {
+    const id = req.params.id;
+    const text = req.body.text
+    if (req.params.hasOwnProperty('_id')
```
Comment (spyatnicyn) [inline src/modules/controllers/task.controller.js:50]: Use validation in middleware to validate it

File: src/modules/controllers/task.controller.js [hunk 23]
```diff
@@ -0,0 +1,97 @@
...
+    res.send({ data: result });
+  } catch (err) {
+    res.status(400).send("error");
+  }
+};
+
+const createNewTask = async (req, res) => {
+  try {
+    const text = req.body.text;
+    if (!req.body.hasOwnProperty('text')
```
Comment (spyatnicyn) [inline src/modules/controllers/task.controller.js:17]: Middleware validation

File: package.json [hunk 24]
```diff
@@ -0,0 +1,20 @@
...
+  },
+  "keywords": [],
+  "author": "",
+  "license": "ISC",
+  "dependencies": {
+    "cors": "^2.8.5",
+    "dotenv": "^16.0.1",
+    "express": "^4.18.1",
+    "express-validator": "^6.14.2",
+    "joi": "^17.6.0",
```
Comment (spyatnicyn) [inline package.json:17]: You installed express-validator and joi, but you are using only express-validator. joi still needed?

File: app.js [hunk 25]
```diff
@@ -0,0 +1,23 @@
...
+  try {
+    mongoose.connect(URL, { useNewUrlParser: true, useUnifiedTopology: true });
+    app.listen(PORT, () => {
+      console.log(`Example app listening on port ${PORT}`);
+    });
+  } catch {
+    console.error('Connection not created!');
+  }
+}
+const apiRoutes = require("./src/modules/routes/task");
```
Comment (spyatnicyn) [inline app.js:18]: Import modules in the beginning of the file

File: src/assist/validation.js [hunk 26]
```diff
@@ -0,0 +1,7 @@
+const validationString = (string) => {
```
Comment (spyatnicyn) [inline src/assist/validation.js:1]: If you will use joi or express validation, you can delete this file

File: package.json (hunk 24 above)
Comment (spyatnicyn) [reply package.json:17]: As i can see, you don't use joi. So you can remove it from dependencies

File: src/modules/controllers/task.controller.js [hunk 27]
```diff
@@ -0,0 +1,79 @@
...
+    res.send({ data: saveTask });
+  } catch (err) {
+    res.status(400).send("error");
+  }
+};
+
+const deleteTask = async (req, res, next) => {
+  try {
+    const { id } = req.params;
+      result = await Tasks.deleteOne({ _id: id });
```
Comment (spyatnicyn) [inline src/modules/controllers/task.controller.js:28]: indents

File: src/models/Schema.js [hunk 28]
```diff
@@ -0,0 +1,11 @@
+const mongoose = require('mongoose');
+const { Schema } = mongoose;
```
Comment (spyatnicyn) [inline src/models/Schema.js:2]: leave empty line after imports block

**Pull request merged**